src/compas/datastructures/assembly/assembly.py

```diff
@@ -0,0 +1,146 @@
+
+    @property
+    def data(self):
+        """dict : A data dict representing the assembly data structure for serialization.
+        """
+        data = {
+            'attributes': self.attributes,
+            'graph': self.graph.data,
+        }
+        return data
+
+    @data.setter
+    def data(self, data):
+        self.attributes.update(data['attributes'] or {})
+        self.graph.data = data['graph']
+
+    def add_part(self, part, key=None, **kwargs):
+        """Add a part to the assembly.
+
+        Parameters
+        ----------
+        part: :class:`compas.datastructures.Node`
+            The part or assembly of parts to add.
+        key: int or str, optional
+            The identifier of the part in the assembly.
+            Note that the key is unique only in the context of the current assembly.
+            Nested assemblies may have the same ``key`` value for one of their parts.
+            Default is ``None`` in which case the key will be automatically assigned integer value.
+        kwargs: dict
+            Additional named parameters colledted in a dict.
+
+        Returns
+        -------
+        int or str
+            The identifier of the part in the current assembly graph.
+
+        """
+        key = self.graph.add_node(key=key, part=part, **kwargs)
+        part.key = key
+        return key
+
+    def add_connection(self, a, b, **kwargs):
+        """Add a connection between two parts.
+
+        Parameters
+        ----------
+        a: int or str
+            The identifier of the "from" part in the current assembly.
+        b: int or str
+            The identifier of the "to" part in the current assembly.
+        kwargs: dict
+            Additional named parameters colledted in a dict.
+
+        Returns
+        -------
+        tuple of str or int
+            The tuple of node identifiers that identifies the connection.
+
+        """
+        return self.graph.add_edge(a, b, **kwargs)
+
+    def parts(self):
+        """The parts of the assembly.
+
+        Yields
+        ------
+        :class:`compas.datastructures.Node`
+            The individual parts or sub-assemblies of the current assembly.
+        """
+        for node in self.graph.nodes():
+            yield self.graph.node_attribute(node, 'part')
+
+    def connections(self):
+        """The connections between the parts."""
+        return self.graph.edges()
+
+    def find(self, guid):
+        """Find a part in the assembly.
+
+        This method will traverse all parts and sub-assemblies of the current assembly
+        to find the part.
+
+        Parameters
+        ----------
+        guid: str
+            A globally unique identifier.
+            This identifier is automatically assigned when parts are created.
+
+        Returns
+        -------
+        :class:`compas.datastructures.Node` or None
+            The identified part, if any.
+
+        """
+        for node in self.graph.nodes():
```

Comment:
`find` by guid should be a O(1) operation, ie. we should maintain a map of parts keyed by guid and just do a lookup. Sub-assemblies will need some extra love like looping over them (which triggers the question of how to get them without having to loop through all parts), and calling the O(1) `.find()`  on them.

Another consideration here is naming. Other libraries and languages use `find` for queries that return 0..n elements, and `get` for lookups that must return 0..1 only. Unless this would create inconsistencies with other parts of compas, I would vote for going a similar route.

Reply:
if we use the guid to make a map, all parts should be findable from the top level since the guid is "guaranteed" to be unique. would not be too complicated to manage if we only allow parts to be added using a method of the assembly and not the (sub)part.